Integrate Centralized BLE Discovery via dbus-ble-advertisements

cgoudie opened the pull request:
## Overview
This PR integrates the SeeLevel service with the `dbus-ble-advertisements` centralized BLE router, eliminating per-service BLE scanning overhead and enabling UI-based sensor control.

## Key Changes

### 1. Centralized BLE Discovery
- **Added**: `data/ble_scanner.py` - Pluggable scanner that subscribes to D-Bus signals from `dbus-ble-advertisements`
- **Removed**: `data/dbus-seelevel-discover.py` - Standalone BLE scanning script no longer needed
- **Benefit**: Reduces Bluetooth stack load, improves discovery speed, eliminates duplicate scanning

### 2. UI-Controlled Sensor Toggles
- Each detected sensor now appears as a `SwitchableOutput` in Venus OS GUI
- Users can enable/disable individual sensors via Switches menu
- Sensor metadata (name, type, capacity) persisted in JSON file
- Added proper D-Bus paths (`FirmwareVersion`, `HardwareVersion`) for GUI display

### 3. Installation & Persistence
- `install.sh` auto-detects `dbus-ble-advertisements` prerequisite
- Integrated with `/data/rc.local` for service persistence across reboots
- Service startup includes retry logic for `dbus-ble-advertisements` availability

### 4. Documentation
- Updated README with centralized discovery explanation
- Added configuration screenshots showing:
  - Sensor control switches in Venus OS UI
  - "Show controls" setting for re-enabling hidden switches
- Documented D-Bus naming conventions

### 5. D-Bus Interface
- **Service Name**: `com.victronenergy.switch.ble.advertisements`
- **Interface Name**: `com.techblueprints.ble.Advertisements`

## Testing
- ✅ Deployed and verified on Cerbo GX
- ✅ Confirmed service persistence after reboot
- ✅ Validated sensor detection and UI control

## Branch Info
- **Base**: `6a93050` (Move installation to /data/apps/dbus-seelevel)
- **Commits**: 28 commits implementing full BLE router integration

Commit:
Fix service name sanitization for D-Bus path compatibility

- Replace hyphens with underscores in service_name to comply with D-Bus object path naming
- Ensures service names like 'seelevel' work correctly in paths like /ble_advertisements/seelevel/mfgr/305

Commit:
Integrate pluggable BLE scanner with dbus-ble-advertisements support

- Replace direct btmon usage with pluggable scanner interface
- Update advertisement_callback to use standardized signature (mac, mfg_id, data, rssi, interface, name)
- Use create_scanner() to automatically select best available backend (dbus-ble-advertisements or btmon fallback)
- Register with dbus-ble-advertisements service when available for centralized BLE scanning
- Register both manufacturer IDs: MFG_ID_CYPRESS (305) and MFG_ID_SEELEVEL (3264)
- Register all configured MAC addresses for optimized filtering
- Properly handle device name from advertisements

Commit:
Fix D-Bus mainloop integration for scanner registration

- Add DBusGMainLoop setup in DBusAdvertisementScanner.start()
- Ensures D-Bus can create service objects and register signals
- Fixes RuntimeError about D-Bus connections needing mainloop attachment

Commit:
Fix D-Bus mainloop setup in is_available() check

- Add DBusGMainLoop setup in is_available() before creating D-Bus connection
- Fixes hang when checking service availability before start() is called
- is_available() is called synchronously during create_scanner(), needs mainloop

Commit:
Add btmon scanner backend for fallback support

- Implements BtmonBLEScanner class using btmon subprocess
- Supports multiple manufacturer IDs (unlike Bleak)
- Parses btmon output to extract MAC, manufacturer ID, RSSI, data, interface, and name
- Adds btmon as fallback between D-Bus and Bleak scanners
- Priority: 1) dbus-ble-advertisements, 2) btmon, 3) Bleak
- Enables seelevel to work without router service

Commit:
WIP: Add btmon fallback scanner (asyncio integration issue)

Commit:
Remove fallback scanners - D-Bus router only (use legacy-standalone-btmon branch for standalone)

Commit:
Add dbus-ble-advertisements prerequisite checks and documentation

Commit:
Remove dead code: legacy btmon parsing functions

Commit:
Fix D-Bus object path for GetVersion check (use /ble_advertisements instead of /)

Commit:
Add install.sh with auto-install prompt for dbus-ble-advertisements

New one-command installation script that:
- Checks for dbus-ble-advertisements
- Offers to auto-install router if missing
- Copies all files to correct locations
- Sets up service symlink
- Reminds user to run discovery tool

Commit:
refactor: Remove MAC address filtering, rely on router UI toggles

- Remove MAC address registration from BLE scanner
- Only register for manufacturer IDs (0x0131 and 0x0CC0)
- Router now handles all device filtering via enabled/disabled UI toggles
- Simplifies client service - no need to track or filter by MAC addresses
- All advertisements for enabled devices are automatically routed by the router
- Update service/run to use unbuffered Python output and check for correct router service name

Commit:
Refactor seelevel to use switch-based sensor configuration

- Remove dependency on config files and discovery tool
- Create switch device (com.victronenergy.switch.seelevel_sensors)
- Dynamically discover sensors from BLE advertisements
- Create switchable outputs for each sensor (tank/temperature/battery)
- Default tanks and temperatures to enabled, battery to disabled
- Persist sensor states to /data/apps/dbus-seelevel/sensors.json
- Start sensor processes only for enabled switches
- Update README with new switch-based configuration approach
- Remove references to dbus-seelevel-discover.py tool

Commit:
Refactor seelevel to add SwitchableOutput to each sensor service

- Remove separate switch device (com.victronenergy.switch.seelevel_sensors)
- Add SwitchableOutput/0/ paths to each sensor in dbus-seelevel-sensor.py
- Each tank, temperature, and battery sensor now has its own built-in switch
- Default states: tanks and temperatures enabled, battery disabled
- Simplify main service to only handle discovery and process management
- Remove MAC address suffix from sensor names (e.g., 'Fresh Water' not 'Fresh Water (8D:95:69)')
- Each sensor handles its own enabled state via SwitchableOutput/0/State

Commit:
Make tank Capacity field writeable in UI

- Changed /Capacity path to be writeable with callback
- Added _on_capacity_changed callback to update Remaining when capacity changes
- Always create /Capacity path for tanks (even if 0)
- Simplified update() method to use current Capacity from D-Bus
- Users can now edit tank capacity directly in the UI

Commit:
Make sensor switches control Connected state

- When a sensor switch is disabled, set /Connected to 0 to hide from GUI
- When enabled, set /Connected to 1 to show in GUI
- Fixed subprocess stdout/stderr to inherit from main service for logging
- Added logging import and configuration to sensor process
- Fixed default_enabled to use integer (1/0) instead of boolean for D-Bus compatibility

Commit:
Simplify sensor enable/disable logic per D-Bus documentation

- Removed Connected state management and value clearing
- Per D-Bus docs, when State is 0 (disabled), updates are skipped via update() method
- Device remains visible in UI but doesn't update values when disabled
- Simpler implementation that follows Venus OS conventions

Commit:
refactor: Move sensor persistence from JSON to D-Bus paths

- Removed sensors.json file and related I/O operations
- Store sensor metadata (MAC, type, sensor_num) as D-Bus paths on switch object
- Added SensorMAC, SensorTypeID, SensorNum, SensorType paths to each SwitchableOutput
- Implement _load_discovered_sensors() to rebuild state from D-Bus on startup
- All configuration now managed in D-Bus (Venus OS conventions)
- Updated README to reflect D-Bus-only persistence
- Removed obsolete apk-analysis directory
- Removed obsolete test scripts (ble_enum_b1.py, nus_sniff.py)
- Removed dbus-seelevel-discover.py (no longer needed)

This follows Venus OS best practices where all persistent state is stored
in D-Bus rather than external JSON files.

Commit:
feat: Implement proper D-Bus settings persistence for sensor metadata

- Store all sensor metadata in com.victronenergy.settings (Venus OS standard)
- Sensor list stored at /Settings/Devices/seelevel_monitor/SensorKeys
- Individual sensor metadata at /Settings/Devices/seelevel_monitor/Sensors/{sensor_key}/
- Metadata includes: MAC, TypeID, Num, Name, Type, RelayID, Enabled
- Removed volatile D-Bus paths (SensorMAC, etc.) from switch objects
- Added _save_discovered_sensors() to persist to settings after changes
- Updated _load_discovered_sensors() to read from settings on startup
- Updated README to document settings-based persistence

This follows Venus OS conventions where all persistent configuration is
stored in com.victronenergy.settings and survives service restarts/reboots.

Commit:
feat: Use JSON file persistence for sensor metadata (proper approach)

Venus OS settings service (com.victronenergy.settings) is designed for static,
pre-defined settings managed by SettingsDevice, not for dynamically created
arbitrary settings. The AddSetting/AddSilentSetting methods return error -3
when attempting to create arbitrary nested paths.

JSON file persistence is the standard Venus OS pattern for dynamically
discovered entities. This approach is used by many core Venus OS services
for similar use cases.

Changes:
- Store sensor metadata in /data/apps/dbus-seelevel/sensors.json
- Simple load/save methods for JSON persistence
- Device settings (ClassAndVrmInstance, CustomName) remain in com.victronenergy.settings
- Updated README to document the hybrid approach
- Clean, maintainable solution that follows Venus OS conventions

This resolves the settings persistence issue and provides a robust,
standard solution for dynamic sensor discovery.

Commit:
fix: Add FirmwareVersion and HardwareVersion paths for UI display

The Venus OS GUI was showing '0' for the device name because it expects
FirmwareVersion and HardwareVersion paths to be present on switch devices.
Without these paths, the GUI falls back to showing ProductId (0xFFFF = 65535)
which displays as '0' in some contexts.

Added both paths to match the BLE Router implementation:
- /FirmwareVersion: '1.0.1'
- /HardwareVersion: None (empty array)

This should fix the switches pane header to show 'SeeLevel Monitor' instead of '0'.

Commit:
fix: Match BLE Router D-Bus structure and improve UI display

Fixed the '0' device name issue by matching BLE Router's exact D-Bus structure:

Device-level changes:
- Changed /Mgmt/ProcessName to __file__ (full path)
- Changed /Mgmt/Connection to 'SeeLevel Monitor'
- Removed CustomName from SettingsDevice registration
- Removed writeable=True from /CustomName
- Added /FirmwareVersion and /HardwareVersion paths

Switch structure changes:
- Unified all sensor switches to match relay_0 structure exactly
- Added ValidTypes, Function, ValidFunctions to all switches
- Changed Group from 0 to empty string
- Made Type and ShowUIControl writeable
- Removed PowerOnState (then added back for clarity)
- Changed CustomName to empty string for consistency

UI improvements:
- Renamed 'SeeLevel Config' to '* SeeLevel Discovery' (sorts to top)
- Fixed discovery switch to properly hide/show sensor switches
- All switches now appear under 'SeeLevel Monitor' device (not split)
- Added debug logging for switch state changes

All switches now display correctly under the SeeLevel Monitor device in the GUI.

Commit:
Add rc.local persistence for service across reboots

Commit:
Add retry logic for dbus-ble-advertisements availability at startup

Commit:
Fix service name check to use correct D-Bus name

Commit:
Standardize on com.victronenergy.switch.ble_router for all checks

Commit:
Update to com.techblueprints.BleAdvertisements interface

Commit:
Update to ble_advertisements service name

Commit:
Update D-Bus references for ble-advertisements router

- Service name: com.victronenergy.switch.ble.advertisements
- Interface name: com.techblueprints.ble.Advertisements
- Updated install.sh, service/run, and ble_scanner.py

Aligns with D-Bus naming conventions in dbus-ble-advertisements.

Commit:
Add configuration documentation with screenshots

- Added sensor management screenshot showing SeeLevel switches UI
- Added 'How to Re-enable Hidden Sensors' guide with step-by-step instructions
- Included screenshot showing Show controls setting
- Explains persistent settings and MAC address identification
- Notes that this process applies to other switch-based services

cgoudie commented:
@atillack would you try this branch on your cerbo after installing the dependent https://github.com/TechBlueprints/dbus-ble-advertisements service? This version has UI, filtering, and other improvements, but I can't verify it works with your model of seelevel :).

atillack commented:
@cgoudie Somebody has been busy - I went over the code and it looks good. I'll test it in a little bit :-)

I would recommend to remove the __pycache__ directories. It probably also would be a good idea to put in GH references/links to the `pycryptodome`,  `bleak`, and `victron-ble` repos instead of copying the files. This way, code fixes are immediately incorporated and it won't look as scary to people that their BLE code has a Crypto folder ;-)

Commit:
Remove __pycache__ directories and add .gitignore

- Remove all Python bytecode cache directories (__pycache__)
- Add comprehensive .gitignore for Python projects:
  - Python bytecode and cache files
  - Virtual environments
  - IDE files
  - OS-specific files
  - Logs and local config

Commit:
Add DEPENDENCIES.md documenting bundled libraries

- Explain why dependencies are bundled (Venus OS constraints)
- Document all bundled libraries with GitHub links:
  - bleak (BLE communication)
  - pycryptodome (cryptography)
  - victron-ble (Victron BLE protocol)
  - velib_python (Venus OS D-Bus integration)
- Provide update instructions for each library
- Reference Victron's dbus-serialbattery as precedent

Addresses reviewer feedback about documenting external code sources.

Commit:
Remove unused dependencies (bleak, victron_ble, pycryptodome)

SeeLevel service architecture uses dbus-ble-advertisements router for BLE
scanning and parses raw manufacturer data for SeeLevel protocol. It does not:
- Use Bleak for direct BLE communication
- Parse Victron BLE encrypted advertisements
- Require cryptographic libraries

Only velib_python is needed for Venus OS D-Bus integration.

Updated DEPENDENCIES.md to reflect minimal dependencies and document
the service architecture.

cgoudie commented:
> @cgoudie Somebody has been busy - I went over the code and it looks good. I'll test it in a little bit :-)
> 
> I would recommend to remove the **pycache** directories. It probably also would be a good idea to put in GH references/links to the `pycryptodome`, `bleak`, and `victron-ble` repos instead of copying the files. This way, code fixes are immediately incorporated and it won't look as scary to people that their BLE code has a Crypto folder ;-)

The other examples I have bundles dependencies in ext/ eg dbus-serialbattery or dbus-aggregate-batteries. I added a DEPENDENCIES.md files with GitHub links, licenses, and update instructions for the bundled libraries though. I also found that I had forgotten to remove some libraries that I included from previous iterations. Appreciate the catch to do so.

Pycache... I was sure that was in my gitignore. Thanks for the catch.

atillack commented:
if it helps here is the BLE advertisement log output:
```
@40000000691a264102f0b944 2025-11-16 19:29:59,048 - Found 0 likely client services to check: []
@40000000691a2641035c0604 2025-11-16 19:29:59,049 - Initial scan complete - found 0 mfgr registrations, 0 MAC registrations
@40000000691a2641035c30fc 2025-11-16 19:29:59,050 - === INITIAL SERVICE SCAN COMPLETE - btmon processing should resume ===
@40000000691a2641035d36b4 2025-11-16 19:29:59,051 - Router service running...
@40000000691a2706248111e4 2025-11-16 19:33:16,611 - New device discovery changed to: True
@40000000691a270625876df4 2025-11-16 19:33:16,622 - Discovery enabled - showing enabled device switches
@40000000691a2722264c63d4 2025-11-16 19:33:44,641 - New device discovery changed to: False
@40000000691a2722271ed6d4 2025-11-16 19:33:44,654 - Discovery disabled - hiding all device switches
@40000000691a28c01937391c 2025-11-16 19:40:38,421 - New device discovery changed to: True
@40000000691a28c01b38d57c 2025-11-16 19:40:38,444 - Discovery enabled - showing enabled device switches
```

cgoudie commented:
> The strange thing is even with only BLE adv I don't see any of the many bluetooth devices around (two Mopeka ones, the Seelevel, many Victron thingies)

I found the bug. I did a last minute rename and missed one of the places. Try now

atillack commented:
Alright, BLE adv. now sees the device but the Seelevel monitor not yet:
<img width="2044" height="1198" alt="image" src="https://github.com/user-attachments/assets/867d57f6-72ab-45fd-941c-37182e43f352" />

Commit:
Make service resilient to router unavailability

Changes:
- Remove router availability check from service/run script
- Service now starts immediately regardless of router state
- BLE scanner creates registrations optimistically
- If router unavailable, service continues without BLE scanning
- Graceful degradation instead of hard failure

This allows:
- Services to start before router
- Services to continue if router crashes/restarts
- Router to discover already-running services on startup
- Independent service lifecycle management

Commit:
Add comprehensive logging for BTP7 sensor discovery

Changes:
- Log all incoming SeeLevel advertisements (debug level)
- Log BTP7 sensor discovery with sensor number and value
- Log BTP7 battery sensor discovery with voltage
- Consistent logging between BTP3 and BTP7 paths

This helps debug why sensors aren't appearing after device discovery.

cgoudie commented:
Or... I'm afraid I've changed you into an incidental beta tester when all I was going to test by you was the device discovery and parsing. I'll reset the objects in dbus so that I'm in a clean slate and rerun. It just might be a bit as I'm about to head for a work trip.

Commit:
Simplify installation: fix path typo and use ln -sf without root remount

- Fix incorrect path in README: /data/apps/dbus-seelevel/data/dbus-seelevel-*.py (was nested incorrectly)
- Simplify install.sh to use 'ln -sf' directly in rc.local instead of install-reboot.sh wrapper
- No root filesystem remount required - all changes are in /data/ which is always writable
- Update README to reflect simplified installation process

cgoudie commented:
Also @atillack ... what is the 3 bars on this?

<img width="141" height="410" alt="image" src="https://github.com/user-attachments/assets/6015bb0d-a60e-4a84-bf09-c9d73014e90a" />

Is it 3 tanks or they just hold that much water?
